src/data/nat/multiplicity.lean

```diff
@@ -0,0 +1,189 @@
+by induction n; simp [nat.pow_succ, hp.multiplicity_mul, *, hp.multiplicity_one, succ_smul]
+
+lemma multiplicity_self {p : ℕ} (hp : p.prime) : multiplicity p p = 1 :=
+have h₁ : ¬ is_unit (p : ℤ), from mt is_unit_int.1 (ne_of_gt hp.one_lt),
+have h₂ : (p : ℤ) ≠ 0, from int.coe_nat_ne_zero.2 hp.ne_zero,
+by rw [int.coe_nat_multiplicity, multiplicity_self h₁ h₂]
+
+lemma multiplicity_pow_self {p n : ℕ} (hp : p.prime) : multiplicity p (p ^ n) = n :=
+by induction n; simp [hp.multiplicity_one, nat.pow_succ, hp.multiplicity_mul, *,
+  hp.multiplicity_self, succ_eq_add_one]
+
+/-- The multiplicity of a prime in `fact n` is the sum of the quotients `n / p ^ i`.
+  This sum is expressed over the set `Ico 1 b` where `b` is any bound at least `n` -/
+lemma multiplicity_fact {p : ℕ} (hp : p.prime) :
+  ∀ {n b : ℕ}, n ≤ b → multiplicity p n.fact = ((Ico 1 b).sum (λ i, n / p ^ i) : ℕ)
+| 0     b hb := by simp [Ico, hp.multiplicity_one]
+| (n+1) b hb :=
+  calc multiplicity p (n+1).fact = multiplicity p n.fact + multiplicity p (n+1) :
+    by rw [fact_succ, hp.multiplicity_mul, add_comm]
+  ... = ((Ico 1 b).sum (λ i, n / p ^ i) : ℕ) + ((finset.Ico 1 b).filter (λ i, p ^ i ∣ n+1)).card :
+    by rw [multiplicity_fact (le_of_succ_le hb),
+      ← multiplicity_eq_card_pow_dvd (ne_of_gt hp.one_lt) (succ_pos _) hb]
+  ... = ((Ico 1 b).sum (λ i, n / p ^ i + if p^i ∣ n+1 then 1 else 0) : ℕ) :
+    by rw [sum_add_distrib, sum_ite (λ _, 1) (λ _, 0) (λ x, x)]; simp
+  ... = ((Ico 1 b).sum (λ i, (n + 1) / p ^ i) : ℕ) :
+    congr_arg coe $ finset.sum_congr rfl (by intros; simp [nat.succ_div]; congr)
+
+/-- A prime power divides `fact n` iff it is at most the sum of the quotients `n / p ^ i`.
+  This sum is expressed over the set `Ico 1 b` where `b` is any bound at least `n` -/
+lemma pow_dvd_fact_iff {p : ℕ} {n r b : ℕ} (hp : p.prime) (hbn : n ≤ b) :
+   p ^ r ∣ fact n ↔ r ≤ (Ico 1 b).sum (λ i, n / p ^ i) :=
+by rw [← enat.coe_le_coe, ← hp.multiplicity_fact hbn, ← pow_dvd_iff_le_multiplicity, nat.pow_eq_pow]
+
+lemma multiplicity_choose_aux {p n b k : ℕ} (hp : p.prime) (hkn : k ≤ n) :
+  (finset.Ico 1 b).sum (λ i, n / p ^ i) =
+  (finset.Ico 1 b).sum (λ i, k / p ^ i) + (finset.Ico 1 b).sum (λ i, (n - k) / p ^ i) +
+  ((finset.Ico 1 b).filter (λ i, p ^ i ≤ k % p ^ i + (n - k) % p ^ i)).card :=
+calc (finset.Ico 1 b).sum (λ i, n / p ^ i)
+    = (finset.Ico 1 b).sum (λ i, (k + (n - k)) / p ^ i) :
+    by simp only [nat.add_sub_cancel' hkn]
+... = (finset.Ico 1 b).sum (λ i, k / p ^ i + (n - k) / p ^ i +
+      if p ^ i ≤ k % p ^ i + (n - k) % p ^ i then 1 else 0) : by simp only [nat.add_div (nat.pow_pos hp.pos _)]
+... = _ : by simp [sum_add_distrib, sum_ite (λ _, 1) (λ _, 0) (λ x, x)]
+
+/-- The multiplity of `p` in `choose n k` is the number of carries when `k` and `n - k`
+  are added in base `p`. The set is expressed by filtering `Ico 1 b` where `b`
+  is any bound at least `n`. -/
+lemma multiplicity_choose {p n k b : ℕ} (hp : p.prime) (hkn : k ≤ n) (hnb : n ≤ b) :
```

Comment:
Nice! I remember discussing this lemma when working on Witt vectors. Back then I found a way around it. But it's cool that this lemma is now available.

Reply:
Which form of the theorem did you want for that? This particular version looks horrible to use to me, so I proved a couple of other corollaries which I thought would be more useful. I did some of my number theory coursework in Lean, and I used the form `multiplicity p n ≤ multiplicity p (choose n k) + multiplicity p k`

Reply:
Nah, I needed exactly the version with carries in the addition in base `p`.